lib/codegen/src/simple_gvn.rs

```diff
@@ -90,18 +99,23 @@ pub fn do_simple_gvn(func: &mut Function, domtree: &mut DominatorTree) {
 
             let func = Ref::map(pos.borrow(), |pos| &pos.func);
 
-            let opcode = func.dfg[inst].opcode();
+            let inst_data = func.dfg[inst].clone();
+            let opcode = inst_data.opcode();
+
             if opcode.is_branch() && !opcode.is_terminator() {
                 scope_stack.push(func.layout.next_inst(inst).unwrap());
                 visible_values.increment_depth();
             }
+
             if trivially_unsafe_for_gvn(opcode) {
                 continue;
+            } else if is_load_and_not_readonly(&inst_data) {
+                continue;
             }
```

Comment:
This pattern is easier to read of we omit the else because the previous if ends with continue. And also, splitting up the two ifs means we can move the inst_data clone() down after the first if. This is a dubious micro-optimization, however I'm expecting we'll proceed to put more code here and I want to keep "check the opcode" as the clearly independent first step, because that quickly eliminates a lot of instructions from further consideration.